feat!: update private_mutable replace functions to closure style (#7065)

wildjos opened the pull request:
# Motivation
Previously, updating a note required reading it first via `get_note` (which nullified and recreated it) and then calling `replace` — effectively proving the note twice.  Now, `replace` accepts a callback that transforms the current note directly, and `initialize_or_replace` uses this updated `replace` internally.  This reduces circuit cost while maintaining exactly one current note.

# Key points

1. `replace(self, new_note)` (old) → `replace(self, f)` (new), where `f` takes the current note and returns a transformed note.
2. `initialize_or_replace(self, note)` (old) → `initialize_or_replace(self, init_note, f)` (new). Uninitialized variables use `init_note`, initialized ones pass the transform function to `replace`.
3. The previous note is automatically nullified before the new note is inserted.
4. `NoteEmission<Note>` still requires `.emit()` or `.discard()`.
5. Example contracts and docs have been updated to reflect this new usage pattern.

# Breaking change
Code relying on the old `get_note() + replace()` pattern may need to be updated.

# Addresses
Closes #7065

olehmisar commented:
Why not `initialize_or_replace<N>(f: fn(prev: Option<N>) -> N)`?

Seems more logical

benesjan commented on docs/docs/migration_notes.md:
```diff
@@ -53,9 +82,9 @@ The note emission API has been significantly reworked to provide clearer semanti
-   - `UNCONSTRAINED_OFFCHAIN`: For offchain delivery (e.g. cloud storage) without constraints. Lowest cost since no onchain storage needed. Requires custom infrastructure for delivery. No guarantees that messages will be delivered or that recipients will ever find them.
+   - `CONSTRAINED_ONCHAIN`: For on-chain delivery with cryptographic guarantees that recipients can discover and decrypt messages. Uses constrained encryption but is slower to prove. Best for critical messages that contracts need to verify.
+   - `UNCONSTRAINED_ONCHAIN`: For on-chain delivery without encryption constraints. Faster proving but trusts the sender. Good when the sender is incentivized to perform encryption correctly (e.g. they are buying something and will only get it if the recipient sees the note). No guarantees that recipients will be able to find or decrypt messages.
+   - `UNCONSTRAINED_OFFCHAIN`: For off-chain delivery (e.g. cloud storage) without constraints. Lowest cost since no on-chain storage needed. Requires custom infrastructure for delivery. No guarantees that messages will be delivered or that recipients will ever find them.
```
[Here](https://aztecprotocol.slack.com/archives/C07R3GTKJ49/p1757430150400379) we agreed that we would go with "offchain" and "onchain" so please revert these changes.

This is also the reason why the docs build is failing:

<img width="766" height="238" alt="image" src="https://github.com/user-attachments/assets/142734a1-045e-4566-a03d-5fd75813e62e" />

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -441,7 +447,7 @@ where
     ///
     /// This function is only available in a UtilityContext (unconstrained environment)
-    /// and is typically used for offchain queries, view functions, or testing.
+    /// and is typically used for off-chain queries, view functions, or testing.
```
This will also make the cspell extension complain.

benesjan commented on noir-projects/noir-contracts/contracts/docs/docs_example_contract/src/main.nr:
```diff
@@ -119,21 +119,14 @@ pub contract DocsExample {
     fn increase_legendary_points() {
-        // Ensure `points` > current value
-        // Also serves as a e2e test that you can `get_note()` and then `replace()`
-        // docs:start:state_vars-PrivateMutableGet
```
These ugly tags are used during docs build to insert the relevant piece of code into the docs. I checked and I see that this one is actually used so you will need to fix this:
<img width="1175" height="246" alt="image" src="https://github.com/user-attachments/assets/2f58fab3-3a08-4430-966a-2ef858bfa5cf" />

Given that you no longer call the `get_note` method here you will need to find another place in the codebase where we do that, place the tag there and if it's in a different file update the path from the screenshot.

wildjos replied:
Ok - have added a separate function `read_legendary_points()` to demonstrate `get_note()` independently of `replace()`, and the tag `state_vars-PrivateMutableGet` points to this new function.

benesjan commented on noir-projects/noir-contracts/contracts/docs/docs_example_contract/src/main.nr:
```diff
@@ -119,21 +119,14 @@ pub contract DocsExample {
+                CardNote::new(points, context.msg_sender())
+            })
+            .emit(&mut context, context.msg_sender(), MessageDelivery.CONSTRAINED_ONCHAIN);
     }
```
```suggestion
        // docs:end:state_vars-PrivateMutableReplace
    }
```
We need to keep this here for docs to build

benesjan commented on noir-projects/noir-contracts/contracts/app/app_subscription_contract/src/main.nr:
```diff
@@ -155,11 +155,13 @@ pub contract AppSubscription {
+            .subscriptions
+            .at(subscriber)
+            .initialize_or_replace(subscription_note, |_old| {
+                SubscriptionNote::new(subscriber, expiry_block_number, tx_count)
```
```suggestion
                subscription_note
```
I think it's preferable to just directly use here the subscription note given that it's already created.

benesjan commented on noir-projects/noir-contracts/contracts/app/app_subscription_contract/src/main.nr:
```diff
@@ -79,16 +79,12 @@ pub contract AppSubscription {
-            user_address,
-            MessageDelivery.CONSTRAINED_ONCHAIN,
-        );
+        let updated_note = storage.subscriptions.at(user_address).replace(|mut note| {
```
Just a detail but the name would make me think that it's dome kind of note type and not a `NoteEmission` type. Would rename this to something like `emission`.

benesjan commented on docs/docs/migration_notes.md:
```diff
@@ -45,6 +45,35 @@ As we prepare for a bigger `Wallet` interface refactor and the upcoming `WalletS
+### PrivateMutable: replace / initialize_or_replace behaviour change
+
+**Motivation:**
+Updating a note used to require reading it first (via `get_note`, which nullifies and recreates it) and then calling `replace` — effectively proving the note twice. Now, `replace` accepts a callback that transforms the current note directly, and `initialize_or_replace` simply uses this updated `replace` internally. This reduces circuit cost while maintaining exactly one current note.
```
```suggestion
Updating a note used to require reading it first (via `get_note`, which nullifies and recreates it) and then calling `replace` — effectively proving a note twice. Now, `replace` accepts a callback that transforms the current note directly, and `initialize_or_replace` simply uses this updated `replace` internally. This reduces circuit cost while maintaining exactly one current note.
```

Being pedantic here but having there "proving a" instead of "proving the" is more precise because the first `get_note` call "re-inserts" the original note and then replace would insert the new one so "the" there pointing to the same note was incorrect.


Also note that another advantage of the new approach is that it allows us to insert more notes in a tx that uses this functionality because we have a `MAX_NOTE_HASHES_PER_TX` constant that limits the amount of notes that can be emitted within a tx. So even though we don't incur additional DA cost related to the note hash being emitted in Ethereum blobs because the re-inserted note gets squashed [1] this change makes us less likely to hit the limit which can be a big win for complex txs.

LMK if something from what I wrote is not clear.

[1] Kernel circuits remove it from the final tx effects because they figure out that the note was inserted and then nullified.

benesjan reviewed (approved):
Very nice work! Love how thorough you were with the docs update.

Feel free to merge once you address my comments 😊

benesjan commented:
> Why not initialize_or_replace<N>(f: fn(prev: Option<N>) -> N)?

@olehmisar I like it 👍 

But I am not sure if it makes sense to tackle because I am not sure what is the future of the `initialize_or_replace` function given that it's used only in 1 contract that is quite stale.

@nventuro WDYT?

But anyway seems fine to tackle this in a followup PR if we decide to do it.

nventuro commented:
>But I am not sure if it makes sense to tackle because I am not sure what is the future of the initialize_or_replace function given that it's used only in 1 contract that is quite stale.

We will definitely continue to use this function. Consider any `Map<T, PrivateMutable>` - you'd need to be able to handle initialization on the fly.

>Why not `initialize_or_replace<N>(f: fn(prev: Option<N>) -> N)`?

This is what I considered initially, but dropped it in favor of @wildjos's suggestion to have the `init_note` and the `transform_fn` separate. My reasoning was that most callsites would end up looking like this

```noir
state_var.initialize_or_replace(|curr| 
 if curr.is_none() {
   MyNote::new(init)
 } else {
   MyNote::new(f(curr.unwrap()))
 }
);
```

and it seemed simpler to instead go with 


```noir
state_var.initialize_or_replace(MyNote::new(init), |curr| MyNote::new(f(curr)));
```

However, it is true that by having an explicit init note param we'd force the caller to create the init note _even if it is not used_. In circuits this would typically not be an issue because you'd need to pay for the predicated constraints anyway, but we also need to consider the possibility of dynamic costs associated to init note creation, such as kernel read requests, external calls, emitted data, etc. So forcing these to be paid is not fantastic. Additionally, the `Option` variant only looks not very nice because we lack proper `enum` support and `match` - if we had that then this'd be:

```noir
state_var.initialize_or_replace(|curr| match curr
  None => MyNote::new(init),
  Some(curr) new(f(curr)),
);
```

So all in all I think the `Option` approach is better - the callback function provides more flexibility to the caller in how they want to construct the init value.

nventuro commented on docs/docs/migration_notes.md:
````diff
@@ -45,6 +45,36 @@ As we prepare for a bigger `Wallet` interface refactor and the upcoming `WalletS
+```diff
+- storage.my_var.replace(new_note);
++ let x: Field = 5;
++ storage.my_var.replace(|note| MockNote::new(note.value + x).build_note();
````
```suggestion
+ storage.my_var.replace(|note| UintNote::new(note.value + x));
```

nventuro replied:
`MockNote` is an internal testing type, that's why it has things like `build_note`. Let's use `UintNote` in these examples, as this is an actual library type that we export.

nventuro commented on docs/docs/migration_notes.md:
````diff
@@ -45,6 +45,36 @@ As we prepare for a bigger `Wallet` interface refactor and the upcoming `WalletS
+
+```diff
+- storage.my_var.initialize_or_replace(init_note);
++ storage.my_var.initialize_or_replace(init_note, |note| MockNote::new(note.value + 5).build_note());
````
```suggestion
+ storage.my_var.initialize_or_replace(init_note, |note| UintNote::new(note.value + 5));
```

nventuro commented on docs/versioned_docs/version-v1.2.0/developers/reference/smart_contract_reference/aztec-nr/aztec/state_vars/private_mutable.md:
````diff
@@ -49,26 +49,27 @@ initialize(self, note);
 
 ```rust
-replace(self, new_note);
+replace(self, f);
````
The `docs` directory contains both a `docs` subdir, with the docs content, and `versioned_docs` (this), with snapshots of the docsite content at older versions. We don't want to update these since `replace` does not look like that in `1.2.0`.

wildjos replied:
Reverted

nventuro commented on noir-projects/noir-contracts/contracts/test/state_vars_contract/src/main.nr:
```diff
@@ -121,32 +121,22 @@ pub contract StateVars {
+            .replace(|old_note| {
+                let new_value = old_note.value() + 1;
+                ValueNote::new(new_value, context.msg_sender())
+            })
```
Lovely

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable/test.nr:
```diff
@@ -122,10 +179,13 @@ unconstrained fn initialize_or_replace_uninitialized() {
 
-        assert_eq(emission.note, note);
+        let emission = state_var.initialize_or_replace(init_note, |old_note: MockNote| {
+            MockNote::new(old_note.value + 1).build_note() // This closure won't be called
```
```suggestion
            assert(false) // This closure won't be called
```

This way we're guaranteeing the closure is not called

nventuro reviewed (commented):
Lovely, thanks a lot! Left some minor comments and replied to the `init + transform` vs `option` question.

wildjos committed:
feat!: update private_mutable replace functions to closure style (#7065)

# Motivation
Previously, updating a note required reading it first via `get_note` (which nullified and recreated it) and then calling `replace` — effectively proving the note twice.  Now, `replace` accepts a callback that transforms the current note directly, and `initialize_or_replace` uses this updated `replace` internally.  This reduces circuit cost while maintaining exactly one current note.

# Key points

1. `replace(self, new_note)` (old) → `replace(self, f)` (new), where `f` takes the current note and returns a transformed note.
2. `initialize_or_replace(self, note)` (old) → `initialize_or_replace(self, init_note, f)` (new). Uninitialized variables use `init_note`, initialized ones pass the transform function to `replace`.
3. The previous note is automatically nullified before the new note is inserted.
4. `NoteEmission<Note>` still requires `.emit()` or `.discard()`.
5. Example contracts and docs have been updated to reflect this new usage pattern.

# Breaking change
Code relying on the old `get_note() + replace()` pattern may need to be updated.

# Addresses
Closes #7065

olehmisar commented:
> > Why not `initialize_or_replace<N>(f: fn(prev: Option<N>) -> N)`?
> 
> This is what I considered initially, but dropped it in favor of @wildjos's suggestion to have the `init_note` and the `transform_fn` separate. My reasoning was that most callsites would end up looking like this
> 
> ```nr
> state_var.initialize_or_replace(|curr| 
>  if curr.is_none() {
>    MyNote::new(init)
>  } else {
>    MyNote::new(f(curr.unwrap()))
>  }
> );
> ```

Can't this be rewritten as:
```rs
state_var.initialize_or_replace(|curr| MyNote::new(curr.unwrap_or(init)))
```

olehmisar commented:
e.g., for a counter:
```rs
my_counter.initialize_or_replace(|x| x.unwrap_or(UintNote::new(0)) + 1)
```

nventuro commented:
>my_counter.initialize_or_replace(|x| x.unwrap_or(UintNote::new(0)) + 1)

Yes, for the simple counter case you could use `unwrap_or`. For the conditional expensive initialization you could use `unwrap_or_else`. If you have some more complicated logic you could presumably combine `map` with those to achieve your goals. So maybe not yet having `match` would not even be such a large concern. Thanks for pointing that out.
